### PR TITLE
Add a transaction history chart BlockScoutWeb

### LINKED ISSUE #811: Add the option for Transactions per day graph on the homepage
Not all networks will have a coin that has a USD value or may just be a testnet or private network. In these cases the homepage graph does not make sense. Instead of the USD chart we should provide a graph showing transactions per minute. All networks will be able to show this. A configurable option in `block_scout_web` would be convenient as well :+1: 

In this same issue, we should also remove the requirement to define a coin in `apps/explorer/config/config.exs`

### PR DESCRIPTION
**Pull request by SlightlyCyborg**
closes #811 

### Enhancements
In Explorer, the market history GenServer was generalized to use the Historian behavior. Added a Transaction.History.Historian to log TransactionStats. 

In Blockscout Web, I show a realtime chart of the TransactionStats.

### Incompatible Changes
* TransactionStats ecto migration required
* Market History uses a historian now
* market_history_chart.js -> history_chart.js

### CODE REVIEW

**Comment by SlightlyCyborg** [comment]
Ok. I'll fix it.

**Comment by SlightlyCyborg** [comment]
Many of the errors in `test_<...>` aren't even on files I changed. There were several tests that were failing when I originally checked out the code.

**Comment by vbaranov** [comment]
@SlightlyCyborg 

>Many of the errors in test_<...> aren't even on files I changed. There were several tests that were failing when I originally checked out the code.

There are some blinking tests in `test_`. It can be fixed by restarting the job in Circle CI. But other tests should not be failing at all. As you can see master branch is not failing.

**Comment by SlightlyCyborg** [comment]
Alright @vbaranov. I made all tests pass.

**Comment by SlightlyCyborg** [comment]
So commit 23019ba passed all tests, but this last commit failed for ethereum_jsonrpc on test_parity_http_websocket, which I assume is one of the blinking tests you were talking about.

**Comment by vbaranov** [comment]
@SlightlyCyborg thank you! we will review your PR soon.

**Comment by SlightlyCyborg** [comment]
@vbaranov 

1) [This](https://github.com/poanetwork/blockscout/pull/2404/files#diff-8550ef3314d4ac11a1138a5143487f25R54) toggles the Explorer process that computes the statistics

[This](https://github.com/poanetwork/blockscout/pull/2404/files#diff-4e67b9d34ff4a9af2e137716b58c6c96R61) toggles which charts are shown in BlockScoutWeb. 

2) On bootup, the Explorer app will compile 365 days worth of transaction statistics by joining  blocks table with transactions table and then counting them. The number of days to fetch data for on bootup is [not configurable](https://github.com/poanetwork/blockscout/pull/2404/files#diff-59096588f9bb792be3291fe879d367efR50). This happens at the very start of bootup, because that is how the Market History was implemented. For testing purposes, this is deactivated. If you find this takes took long when rebooting prod, I can make the historian check the database to see if statistics were already saved before selecting all the transactions and counting them.

3) The process runs every 60 minutes and can be changed with [this](https://github.com/poanetwork/blockscout/pull/2404/files#diff-59096588f9bb792be3291fe879d367efR50) env var (:Explorer :history_fetch_interval). The only way to know if the 20% threshold is exceeded is to [do the DB select for transactions](https://github.com/poanetwork/blockscout/pull/2404/files#diff-0e35d7b780b58d157db92884d81295acR28) which is the most intensive part of updating the statistics. If we do the most intensive part of updating the statistics every 60 minutes just to decide if the threshold has been crossed, we might as well just go ahead and insert the results into the database regardless of what the threshold check was. For that reason, I didn't implement the thresholding and instead just updated the statistics every 60 minutes.

**Comment by SlightlyCyborg** [comment]
Only Transactions
transaction_only

Transactions and MarketCap/Price
transaction_and_marketcap_price

Only MarketCap/Price
marketcap_price

**Comment by SlightlyCyborg** [comment]
TX only
tx

Both
both

Original (Price and MarketCap)
original

**Comment by vbaranov** [comment]
@SlightlyCyborg 

>This toggles the Explorer process that computes the statistics

Let's add ENV VAR `ENABLE_TXS_STATS` to enable/disable txs stats computing. By default, it should be disabled.

>This toggles which charts are shown in BlockScoutWeb.

I think we need env vars: `SHOW_PRICE_CHART`, `SHOW_TXS_CHART`. By default, they should be switched off. Otherwise, it is impossible to manage it without recompiling.

>The process runs every 60 minutes and can be changed with this env var (:Explorer :history_fetch_interval). The only way to know if the 20% threshold is exceeded is to do the DB select for transactions which is the most intensive part of updating the statistics. If we do the most intensive part of updating the statistics every 60 minutes just to decide if the threshold has been crossed, we might as well just go ahead and insert the results into the database regardless of what the threshold check was. For that reason, I didn't implement the thresholding and instead just updated the statistics every 60 minutes.

Ok, got it. But I didn't find where I can fill `:history_fetch_interval` env var. I think you should add something like this: `System.get_env(TRANSACTIONS_HISTORY_FETCH_INTERVAL)`, should you?

**Comment by SlightlyCyborg** [comment]
> I think we need env vars: SHOW_PRICE_CHART, SHOW_TXS_CHART. By default, they should be switched off.

Maybe SHOW_PRICE_CHART on by default to preserve current behavior without modifying ENV (which isn't tracked by git)?

**Comment by vbaranov** [comment]
>Maybe SHOW_PRICE_CHART on by default to preserve current behavior without modifying ENV (which isn't tracked by git)?

Makes sense

**Comment by SlightlyCyborg** [comment]
> Makes sense

K. I'll do that. The UI isn't set up to gracefully handle no graph data.

broken

**Comment by SlightlyCyborg** [comment]
`ENABLE_TXS_STATS, SHOW_PRICE_CHART, SHOW_TXS_CHART,` have all been added as requested.

Also, I added `HISTORY_FETCH_INTERVAL` instead of `TRANSACTION_HISTORY_FETCH_INTERVAL` because at the moment, the intervals for both market history and transaction history are coupled. I can decouple them if that is a problem.

File: apps/block_scout_web/assets/js/pages/chain.js [hunk 1]
```diff
@@ -107,6 +107,11 @@ function baseReducer (state = initialState, action) {
         })
       }
     }
+    case 'RECIEVED_UPDATED_TRANSACTION_STATS': {
```
Comment (pasqu4le) [inline apps/block_scout_web/assets/js/pages/chain.js]: `RECIEVED` typo

File: apps/block_scout_web/lib/block_scout_web/controllers/chain_controller.ex [hunk 2]
```diff
@@ -74,15 +103,9 @@ defmodule BlockScoutWeb.ChainController do
...
-          paging_options: %PagingOptions{page_size: 4},
-          necessity_by_association: %{
-            [miner: :names] => :optional,
-            :transactions => :optional,
-            :rewards => :optional
-          }
-        ]
+        [paging_options: %PagingOptions{page_size: 4}]
         |> Chain.list_blocks()
+        |> Repo.preload([[miner: :names], :transactions, :rewards])
```
Comment (pasqu4le) [inline apps/block_scout_web/lib/block_scout_web/controllers/chain_controller.ex]: There is no need to change this, the `necessity_by_association`'s `:optional`s do exactly a preload.

File: apps/explorer/config/config.exs [hunk 3]
```diff
@@ -51,7 +51,18 @@ config :explorer, Explorer.KnownTokens, enabled: true, store: :ets
 
 config :explorer, Explorer.Integrations.EctoLogger, query_time_ms_threshold: :timer.seconds(2)
 
-config :explorer, Explorer.Market.History.Cataloger, enabled: true
+config :explorer, Explorer.Market.History.Historian, enabled: true
+
+if System.get_env("HISTORY_FETCH_INTERVAL") do
+  mins = String.to_integer(System.get_env("HISTORY_FETCH_INTERVAL"))
+  config :explorer, Explorer.History.Process, history_fetch_interval: :timer.minutes(mins)
+end
```
Comment (pasqu4le) [inline apps/explorer/config/config.exs]: There is a small problem with this block, in that `String.to_integer` can raise an error if `HISTORY_FETCH_INTERVAL` is set to a string that is not a valid number.

I also see you set the default value to be 60 minutes in the code itself, so I suggest replacing this `if` block with:
```elixir
history_fetch_interval =
  case Integer.parse(System.get_env("HISTORY_FETCH_INTERVAL", "")) do
    {mins, ""} -> mins
    _ -> 60
  end
  |> :timer.minutes()

config :explorer, Explorer.History.Process, history_fetch_interval: history_fetch_interval
```

File: apps/explorer/config/config.exs [hunk 4]
```diff
@@ -51,7 +51,18 @@ config :explorer, Explorer.KnownTokens, enabled: true, store: :ets
...
+
+if System.get_env("ENABLE_TXS_STATS") do
+  config :explorer, Explorer.Chain.Transaction.History.Historian, enabled: true
+else
+  config :explorer, Explorer.Chain.Transaction.History.Historian, enabled: false
+end
```
Comment (pasqu4le) [inline apps/explorer/config/config.exs]: no need for the `if` block here, just go with:
```elixir
config :explorer, Explorer.Chain.Transaction.History.Historian, enabled: System.get_env("ENABLE_TXS_STATS", "false") != "false"
```

File: apps/explorer/lib/explorer/history/process.ex [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  @typep milliseconds :: non_neg_integer()
+
+  @spec config_or_default(atom(), term(), module()) :: term()
+  def config_or_default(key, default, module \\ __MODULE__) do
+    Application.get_env(:explorer, module, [])[key] || default
+  end
+
+  @spec base_backoff :: milliseconds()
+  defp base_backoff do
+    config_or_default(:base_backoff, 100)
```
Comment (pasqu4le) [inline apps/explorer/lib/explorer/history/process.ex]: what is the need for `:base_backoff`? Is it only used for testing?

Comment (SlightlyCyborg) [reply apps/explorer/lib/explorer/history/process.ex]: Ya. It is mostly for testing. Can be used as config parameter later if a historian makes API calls and needs a particular timeout between requests. 

Much of the code from process.ex was taken from `lib/explorer/market/history/cataloger.ex`

**Comment by vbaranov** [comment]
@SlightlyCyborg when you will fix @pasqu4le comments could you please close this PR and open it to this branch https://github.com/poanetwork/blockscout/tree/transactions-history in order to make us able to fix some styles or make some other fixes that indirectly related to txs history feature?

**Comment by SlightlyCyborg** [comment]
Sure @vbaranov. I am doing this for a gitcoin bounty, so would it still close #811 when the PR gets accepted?

**Comment by vbaranov** [comment]
@SlightlyCyborg 

> Sure @vbaranov. I am doing this for a gitcoin bounty, so would it still close #811 when the PR gets accepted?

That is right. When there will be no more reviewers requests for changes to your work we will merge it to `transactions-history` branch and can close the issue. And I will leave the comment there that we will make some cosmetic updates to that feature before publishing it but I suppose, it shouldn't slow down paying for your efforts.

**Pull request closed without merging**